packages/belos/tpetra/src/solvers/Belos_Tpetra_GmresSingleReduce.hpp

```diff
@@ -288,39 +299,56 @@ class GmresSingleReduce : public Gmres<SC, MV, OP> {
           M.apply (R, MP);
           X.update (one, MP, one);
         }
-	else {
+        else {
           MVT::MvTimesMatAddMv (one, *Qj, y, one, X);
         }
         y.resize (restart+1);
       }
 
       // Compute explicit residual vector in preparation for restart.
-      P = * (Q.getVectorNonConst (0));
-      A.apply (X, P);
-      P.update (one, B, -one);
-      r_norm = P.norm2 (); // residual norm
+      A.apply (X, R);
+      R.update (one, B, -one);
+      r_norm = R.norm2 (); // residual norm
       output.absResid = r_norm;
       output.relResid = r_norm / b0_norm;
+      if (outPtr != nullptr) {
+        *outPtr << "Implicit and explicit residual norms at restart: " << r_norm_imp << ", " << r_norm << endl;
+      }
 
       metric = this->getConvergenceMetric (r_norm, b0_norm, input);
       if (metric <= input.tol) {
         output.converged = true;
-	return output;
+        return output;
       }
       else if (output.numIters < input.maxNumIters) {
-        // Initialize starting vector for restart
-        if (input.precoSide == "left") {
-          Tpetra::deep_copy (R, P);
-          M.apply (R, P);
-	  // FIXME (mfh 14 Aug 2018) Didn't we already compute this above?
-          r_norm = P.norm2 ();
+        // Restart, only if max inner-iteration was reached.
+        // Otherwise continue the inner-iteration.
+        if (iter >= restart) {
+          // Initialize starting vector for restart
+          iter = 0;
+          P = * (Q.getVectorNonConst (0));
+          if (input.precoSide == "left") {
+            M.apply (R, P);
+            // FIXME (mfh 14 Aug 2018) Didn't we already compute this above?
+            r_norm = P.norm2 ();
+          }
+          else {
+            // set the starting vector
+            Tpetra::deep_copy (P, R);
+          }
+          P.scale (one / r_norm);
+          y[0] = SC {r_norm};
+          for (int i=1; i < restart+1; i++) {
+            y[i] = zero;
+          }
+          output.numRests++; // restart
+        }
+        else {
+          // reset to the old solution
+          if (outPtr != nullptr) *outPtr << " > not-restart with iter=" << iter << endl;
```

Comment:
Prefer this style:
```c++
if (outPtr != nullptr) {
  *outPtr << " > not-restart with iter=" << iter << endl;
}
```
Using braces with if, for, etc. avoids possible bugs and compiler warnings with nested statements.